Error when creating log forwarding bucket notification

```
╷
│ Error: error putting S3 Bucket Notification Configuration: InvalidArgument: Unable to validate the following destination configurations
│ 	status code: 400, request id: 9N23BKE9Z7V2YPMY, host id: SfoD0D+LCrk36nBn/OqDC50rw9tHwk+BezRWzIgLn43TBANqry1Q03UVZHzFI/KfEJJY2gv13Kg=
│ 
│   with aws_s3_bucket_notification.indexer_edu-ucsc-gi-platform-hca-dev-logs_us-east-1_notify,
│   on api_gateway.tf.json line 802, in resource[7].aws_s3_bucket_notification[0].indexer_edu-ucsc-gi-platform-hca-dev-logs_us-east-1_notify:
│  802:                     }
│ 
╵
make[1]: *** [Makefile:60: auto_apply] Error 1
make[1]: Leaving directory '/builds/ucsc/azul/terraform'
make: *** [Makefile:103: auto_terraform] Error 2
```

Our working hypothesis is that this is caused by a race condition between `aws_s3_bucket_notification.indexer_edu-ucsc-gi-platform-hca-dev-logs_us-east-1_notify` and `aws_lambda_permission.indexer_forward_alb_logs-s3event` in the Chalice-generated Terraform configuration.

https://stackoverflow.com/questions/67010382/terraform-error-putting-s3-notification-configuration-invalidargument-unable, 

- [ ] Security design review completed; the Resolution of this issue does **not** …
  - [ ] … affect authentication; for example:
    - OAuth 2.0 with the application (API or Swagger UI)
    - Authentication of developers with Google Cloud APIs
    - Authentication of developers with AWS APIs
    - Authentication with a GitLab instance in the system
    - Password and 2FA authentication with GitHub
    - API access token authentication with GitHub
    - Authentication with 
  - [ ] … affect the permissions of internal users like access to
    - Cloud resources on AWS and GCP
    - GitLab repositories, projects and groups, administration
    - an EC2 instance via SSH
    - GitHub issues, pull requests, commits, commit statuses, wikis, repositories, organizations
  - [ ] … affect the permissions of external users like access to
    - TDR snapshots
  - [ ] … affect permissions of service or bot accounts
    - Cloud resources on AWS and GCP
  - [ ] … affect audit logging in the system, like
    - adding, removing or changing a log message that represents an auditable event
    - changing the routing of log messages through the system
  - [ ] … affect monitoring of the system
  - [ ] … introduce a new software dependency like
    - Python packages on PYPI
    - Command-line utilities
    - Docker images
    - Terraform providers
  - [ ] … add an interface that exposes sensitive or confidential data at the security boundary
  - [ ] … affect the encryption of data at rest
  - [ ] … require persistence of sensitive or confidential data that might require encryption at rest
  - [ ] … require unencrypted transmission of data within the security boundary
  - [ ] … affect the network security layer; for example by 
    - modifying, adding or removing firewall rules
    - modifying, adding or removing security groups
    - changing or adding a port a service, proxy or load balancer listens on
- [ ] Documentation on any unchecked boxes is provided in comments below

--- a/src/azul/terraform.py
+++ b/src/azul/terraform.py
@@ -1,3 +1,6 @@
+from collections import (
+    defaultdict,
+)
 from collections.abc import (
     Iterable,
     Sequence,
@@ -721,6 +724,19 @@ class Chalice:
                 key.replace('.', '_'): value
                 for key, value in bucket_notifications.items()
             }
+            # To prevent a race condition by Terraform, we make the bucket
+            # notifications depend on the related aws_lambda_permission.
+            permissions_by_function = defaultdict(set)
+            for permission_name, permission in resources['aws_lambda_permission'].items():
+                function_ref = permission['function_name']
+                permissions_by_function[function_ref].add(permission_name)
+            for notification in resources['aws_s3_bucket_notification'].values():
+                assert 'depends_on' not in notification, notification
+                notification['depends_on'] = [
+                    f'aws_lambda_permission.{permission_name}'
+                    for function in notification['lambda_function']
+                    for permission_name in permissions_by_function[function['lambda_function_arn']]
+                ]
 
         # The fix for https://github.com/aws/chalice/issues/1237 introduced the
         # create_before_destroy hack and it may have helped but has far-ranging

